### ISSUE TITLE
Test priorities/ordering

### ISSUE THREAD

**Issue by whot** [mesonbuild/meson#5581]
Right now, the order of tests during `ninja -C build test` is the order they were defined in `meson.build`. In my case, some tests take forever, others are very quick. I know that if one of the quick tests fails, the test suite fill fail so there's no point waiting an hour for the rest of the tests to finish.

Moving the quick tests up in the meson.build file means moving them outside their logical scope. e.g. if I build a helper library that's used by the main code, I want the tests for that helper library defined right there, not in some other part of `meson.build`.

So what I'd like is some way to influence the test order, e.g. by assigning a numerical priority to the test. Tests will then be executed in decreasing priority, with all the tests defaulting to 0.

Is this something worth implementing?

**Comment by QuLogic** [comment]
You could use the `--num-processes` argument to `meson test` to run in parallel, I think.

**Comment by scivision** [comment]
In general there will be many tests and the slow ones can get picked first, despite default testing in parallel. Being able to select test priority and/or dependency is a needed feature, which is available in some detail in Ctest.

**Comment by scivision** [comment]
for example if test A fails, don't try to run test B.

**Comment by jpakkane** [comment]
Featurewise this seems like a useful thing. How would this look in practice? Would there be priority levels between 1-10 with 1 being most important and 10 least, with 5 being the default?

**Comment by scivision** [comment]
I think that test dependency is the most necessary, and possibly the most straightforward to implement.
I would *really* like (and need for certain projects) to have a meson.build snippet like

```meson
big_exe = exectuable(...)
little_validator_exe = exectuable(...)

little_test = test('little', little_validator_exe, ...)
big_test = test('big', big_exe, depends: little_test, ...)
```

where `test(depends: foo)` looks at a boolean property foo.passed


---

CMake test properties reference:
* [FIXTURES](https://cmake.org/cmake/help/latest/prop_test/FIXTURES_REQUIRED.html#prop_test:FIXTURES_REQUIRED) this is how Ctest does dependencies. Probably excessive for Meson.
* [COST](https://cmake.org/cmake/help/latest/prop_test/COST.html) is for ordering (lower priority to implement in my opinion)

**Comment by whot** [comment]
> Featurewise this seems like a useful thing. How would this look in practice? Would there be priority levels between 1-10 with 1 being most important and 10 least, with 5 being the default?

first gut reaction would be: make it centered on 0. positive numbers increase priority, negative ones decrease priority and there's no need to add a fixed range. It's just numbers anyway and it gives anyone the flexibility to pick a priority of -1000 for the least important tests.

question is then what to do with subprojects: if our tests are 0, 10, 500 and a subproject has 0, 1000 do we run them in the implied order? or do we need to associate an offset with the subproject to rebase it into our project?

but i agree with @scivision, implementing test dependencies would largely obsolete the priority order but provide some more useful features on top. at least for my use-case it'd be enough because the initial test everything relies on is quick.

but there is an issue with pure serial dependencies though, similar to CPU branch prediction - tests may only have a pseudo-dependency on each other: if test A failed I don't want to start test B. But if test A is still running and I have CPU cycles to spare, I still want to start test B for the optimistic case that A will succeed.

**Comment by jpakkane** [comment]
> implementing test dependencies would largely obsolete the priority order but provide some more useful features on top

However it means that you have to add that dependency to _every_ other test declaration. Having it with higher priority and bailing out on first error makes it a lot simpler to use.

**Comment by scivision** [comment]
Yes if having an integer priority instead of dependency is easier, that is fine too--whatever would allow some way to not run expensive tests if cheap preliminary tests fails.

**Comment by scivision** [comment]
A workaround for this at least for CI is to specify test name(s) or suite(s) at the command line, so that cheap preliminary tests are run first, and then following `meson test` invocations specify the expensive tests. I didn't realize this was possible until https://github.com/mesonbuild/meson/pull/5679

example in CI:

```yaml
- script: meson test cheap1 cheap2
- script: meson test big1 
```
